pkg/master/manager.go

```diff
@@ -0,0 +1,167 @@
+	return &Manager{
+		nodeName:                     nodeName,
+		client:                       myClient,
+		wasEndpointAccessibleAtStart: false,
+		log:                          ctrl.Log.WithName("master").WithName("Manager"),
+	}
+}
+
+func (manager *Manager) Start(ctx context.Context) error {
+	if err := manager.initializeManager(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (manager *Manager) IsMaster() bool {
+	return manager.nodeRole == peers.Master
+}
+
+func (manager *Manager) IsMasterHealthy(workerPeerResponse peers.Response, isOtherMastersCanBeReached bool) bool {
+	switch workerPeerResponse.Reason {
+	//reported unhealthy by worker peers
+	case peers.UnHealthyBecauseCRFound:
+		return false
+	case peers.UnHealthyBecauseNodeIsIsolated:
+		return isOtherMastersCanBeReached
+	//reported healthy by worker peers
+	case peers.HealthyBecauseErrorsThresholdNotReached, peers.HealthyBecauseCRNotFound:
+		return true
+	//master node has connection to most workers, we assume it's not isolated (or at least that the master node that does not have worker peers quorum will reboot)
+	case peers.HealthyBecauseMostPeersCantAccessAPIServer:
+		return manager.isDiagnosticsPassed()
+	case peers.HealthyBecauseNoPeersWereFound:
+		return manager.isDiagnosticsPassed() && isOtherMastersCanBeReached
+
+	default:
+		manager.log.Error(processError, "node is considered unhealthy by worker peers for an unknown reason", "reason", workerPeerResponse.Reason, "node name", manager.nodeName)
+		return false
+	}
+
+}
+
+func (manager *Manager) isDiagnosticsPassed() bool {
+	if manager.isEndpointAccessLost() {
+		return false
+	} else if !manager.isKubeletServiceRunning() {
+		return false
+	} else if !manager.isEtcdRunning() {
+		return false
+	}
+	return true
+}
+
+func wrapWithInitError(err error) error {
+	return fmt.Errorf(initErrorText+" [%w]", err)
+}
+
+func (manager *Manager) initializeManager() error {
+	nodesList := &corev1.NodeList{}
+	if err := manager.client.List(context.TODO(), nodesList, &client.ListOptions{}); err != nil {
+		manager.log.Error(err, "could not retrieve nodes")
+		return wrapWithInitError(err)
+	}
+	var node corev1.Node
+	for _, n := range nodesList.Items {
+		if n.Name == manager.nodeName {
+			manager.setNodeRole(n)
+			node = n
+			break
+		}
+	}
+
+	if &node == nil {
+		manager.log.Error(initError, "could not find node")
+		return initError
+	}
+
+	manager.wasEndpointAccessibleAtStart = manager.isEndpointAccessible()
+	return nil
+}
+
+func (manager *Manager) setNodeRole(node corev1.Node) {
+	if _, isWorker := node.Labels[peers.WorkerLabelName]; isWorker {
+		manager.nodeRole = peers.Worker
+	} else {
+		peers.SetControlPlaneLabelType(&node)
+		if _, isMaster := node.Labels[peers.GetUsedControlPlaneLabel()]; isMaster {
+			manager.nodeRole = peers.Master
+		}
+	}
+}
+
+func (manager *Manager) isEndpointAccessLost() bool {
+	if !manager.wasEndpointAccessibleAtStart {
```

Comment:
The assumption is that if we had a connection when the operator started and now we don't, it would be a good indication that something went wrong.
This is used to differentiate with the use case that we never had a connection which does not mean something went wrong.
In a nutshell I'm looking for something that used to work but not working anymore.